### PR TITLE
Add EncodeURL function

### PR DESCRIPTION
**Pull request by justinpark0916**
Issue #1002 
This adds `EncodeUrl` functionality. The description of the function is described in [this document](https://learn.microsoft.com/en-us/power-platform/power-fx/reference/function-encode-decode).

The expected values on the test files are from PowerApps' EncodeUrl result and the test is checking if they are providing the identical result.

### CODE REVIEW

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/EncodeURL.txt [hunk 1]
```diff
@@ -0,0 +1,16 @@
+// ******** STRING PARAMETERS ********
+
+>> EncodeUrl("https://make.powerapps.com/")
+"https%3A%2F%2Fmake.powerapps.com%2F"
+
+>> EncodeUrl("powerfx&test")
+"powerfx%26test"
+
+>> EncodeUrl("")
```
Comment (anderson-joyle) [inline src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/EncodeURL.txt:9]: >EncodeUrl("") [](http://example.com/codeflow?start=3&length=13)

Please also add test cases where the function receives as argument:
- Blank
- Error
- Unsupported type (number, date, ...).
- Concatenation e.g. EncondeUrl("www" & "msft.com/")
- Concatenation with coercion e.g. EncondeUrl("www.msft.com/?orderid=" & numberVariable)
- Call other functions within e.g. EncondeUrl(Upper("www.msft.com/"))

Also:
- Use EncondeUrl within With function e.g. With({url:"www.msft.com/"}, EncondeUrl(url))

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs [hunk 2]
```diff
@@ -23,6 +23,7 @@ internal static partial class Library
         /// Texl Instance of <see cref="DistinctFunction"/>.
         /// </summary>
         public static readonly TexlFunction DistinctInterpreterFunction = new DistinctFunction();
+        public static readonly TexlFunction EncodeUrlFunction = new EncodeUrlFunction();
```
Comment (anderson-joyle) [inline src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs:26]: >public static readonly TexlFunction EncodeUrlFunction = new EncodeUrlFunction(); [](http://example.com/codeflow?start=8&length=80)

We can remove this and add a new reference at C:\Users...\Microsoft.PowerFx.Core\Texl\BuiltinFunctionsCore.cs. Something like:
public static readonly TexlFunction IsUTCToday = _featureGateFunctions.Add(new IsUTCTodayFunction());

We might change this from featureGateFunctions to _library after chatting with PA team.

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/EncodeURL.txt (hunk 1 above)
Comment (justinpark0916) [reply src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/EncodeURL.txt:10]: Added more tests. Please let me know if there's more scenario that you want to add :)

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs (hunk 2 above)
Comment (justinpark0916) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs]: Fixed!

File: src/strings/PowerFxResources.en-US.resx [hunk 3]
```diff
@@ -5056,4 +5056,15 @@
...
   </data>
+  <data name="AboutEncodeUrl" xml:space="preserve">
+    <value />
+    <comment>Encodes a URL string, replacing certain non-alphanumeric characters with % and a hexadecimal number.</comment>
+  </data>
+  <data name="EncodeUrlArg1" xml:space="preserve">
+    <value>url</value>
+    <comment>Function_parameter - First parameter for the EncodeUrl function - the URL(string) to encode. Translate this string. Maintain as a single word (do not add spaces)</comment>
+  </data>
+  <data name="Abouturl_text" xml:space="preserve">
```
Comment (anderson-joyle) [inline src/strings/PowerFxResources.en-US.resx:5067]: >Abouturl_text [](http://example.com/codeflow?start=14&length=13)

Please change to AboutEncodeUrl_url

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/EncodeURL.txt [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+// String Paremeter Concatenation With Boolean
+>> EncodeUrl("www.msft.com/?orderid=" & false)
+"www.msft.com%2F%3Forderid%3Dfalse"
+
+// PowerFX Parameter with Upper function
+>> EncodeUrl(Upper("www.msft.com/"))
+"WWW.MSFT.COM%2F"
+
+// PowerFX Parameter with Blank function
+>> EncodeUrl(Blank())
```
Comment (anderson-joyle) [inline src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/EncodeURL.txt:38]: >EncodeUrl(Blank()) [](http://example.com/codeflow?start=3&length=18)

Should this return an empty string or a Blank()? I dont have this answer but Ill find out. This is not a blocker to get this PR merged though.

File: src/libraries/Microsoft.PowerFx.Core/Texl/BuiltinFunctionsCore.cs [hunk 5]
```diff
@@ -79,6 +79,7 @@ internal class BuiltinFunctionsCore
         public static readonly TexlFunction Degrees = _library.Add(new DegreesFunction());
         public static readonly TexlFunction DegreesT = _library.Add(new DegreesTableFunction());
         public static readonly TexlFunction DropColumns = _library.Add(new DropColumnsFunction());
+        public static readonly TexlFunction EncodeUrl = _library.Add(new EncodeUrlFunction());
```
Comment (anderson-joyle) [inline src/libraries/Microsoft.PowerFx.Core/Texl/BuiltinFunctionsCore.cs:82]: >public static readonly TexlFunction EncodeUrl = _library.Add(new EncodeUrlFunction()); [](http://example.com/codeflow?start=8&length=86)

@carlosfigueira How risky would be to add this function to the core list? Can we get the JS implementation?

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/EncodeURL.txt (hunk 1 above)
Comment (anderson-joyle) [reply src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/EncodeURL.txt:10]: One more, actually:
>> EncodeUrl(Error({Kind:ErrorKind.Div0}))
Error({Kind:ErrorKind.Div0})

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/EncodeURL.txt (hunk 4 above)
Comment (CarlosFigueiraMSFT) [reply src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/EncodeURL.txt:38]: Blank inputs in string arguments should be treated as empty string, so empty string is the correct answer here.

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/EncodeURL.txt (hunk 1 above)
Comment (justinpark0916) [reply src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/EncodeURL.txt:10]: Updated! (`EncodeUrl(" ")` already exists and `EncodeUrl("[http://b*()!\'ool](http://b%2A%28%29%21%5C%27ool/)")` should expect http%3A%2F%2Fb%2A%28%29%21%5C%27ool

File: src/strings/PowerFxResources.en-US.resx (hunk 3 above)
Comment (justinpark0916) [reply src/strings/PowerFxResources.en-US.resx]: Changed!

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/EncodeURL.txt [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+"3%2F1%2F2023"
+
+// ErrorKind Parameter
+>> EncodeUrl(Error({Kind:ErrorKind.Div0}))
+Error({Kind:ErrorKind.Div0})
+
+// String Parameter
+>> EncodeUrl("a&b?c d%e")
+"a%26b%3Fc%20d%25e"
+
```
Comment (LucGenetier) [inline src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/EncodeURL.txt:68]: I would try some Unicode characters like U+2764 or U+1F606

File: src/strings/PowerFxResources.en-US.resx [hunk 7]
```diff
@@ -5058,4 +5058,15 @@
     <value>Get the default locale set at runtime.</value>
     <comment>Description of 'Language' function.</comment>
   </data>
+  <data name="AboutEncodeUrl" xml:space="preserve">
+    <value />
```
Comment (jas-valgotar) [inline src/strings/PowerFxResources.en-US.resx:5062]: ><value /> [](http://example.com/codeflow?start=4&length=9)

Was this purposefully kept blank?

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/EncodeURL.txt (hunk 6 above)
Comment (justinpark0916) [reply src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/EncodeURL.txt:68]: Added tests on `EncodeUrl(U+2764)` and `EncodeUrl("U+2764")`

File: src/strings/PowerFxResources.en-US.resx (hunk 7 above)
Comment (justinpark0916) [reply src/strings/PowerFxResources.en-US.resx]: Value and comment fields are updated!

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/libraries/Microsoft.PowerFx.Core/Texl/BuiltinFunctionsCore.cs (hunk 5 above)
Comment (anderson-joyle) [reply src/libraries/Microsoft.PowerFx.Core/Texl/BuiltinFunctionsCore.cs:82]: PA, PFx, and PMs are in sync with this change.

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/EncodeURL.txt (hunk 6 above)
Comment (justinpark0916) [reply src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/EncodeURL.txt:68]: Added!

**Comment by LucGenetier** [comment]
✅ No public API change.

**Review by jas-valgotar — APPROVED** [review]

**Review by anderson-joyle — APPROVED** [review]
:shipit:

**Review by MikeStall — APPROVED** [review]

**Pull request merged**